### PR TITLE
Add Conv3D with introducing AbstractConv

### LINKED ISSUE #79: [Video] Add Conv3D layer
We are missing some layers to support the export of models from Keras fully. One of them is the Conv3D layer. A detailed description of the layer can be found [here](https://keras.io/api/layers/convolution_layers/convolution3d/).

- [ ] add layer class to the  [package](https://github.com/JetBrains/KotlinDL/tree/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional) using Conv2D as a [reference implementation](https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/Conv2D.kt)
- [ ] add documentation
- [ ] add test
- [ ] add export function (see [ModelSaver.kt](https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelSaver.kt)) 
- [ ] add import function (see [ModelLoader.kt](https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelLoader.kt))
- [ ] (optional) create an example

This issue depends on the results of the #59

### PR DESCRIPTION
**Pull request by avan1235**
Resolves #79. Work status:
* [x]  add layer class to the  [package](https://github.com/JetBrains/KotlinDL/tree/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional) using Conv2D as a [reference implementation](https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/Conv2D.kt)
* [x]  add documentation
* [x]  add test
* [x]  add export function (see [ModelSaver.kt](https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelSaver.kt))
* [x]  add import function (see [ModelLoader.kt](https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelLoader.kt))
* [ ]  (optional) create an example

Addtionally, I refactor the conv layers to use common implementation for the logic of computing the sizes of in/out shapes (for now not including separable conv - I would do it in separate PR)

### CODE REVIEW

**Comment by avan1235** [comment]
Please create separate issue for example with conv1 and conv3d so I could focus on this topic after the merge of this change

**Comment by zaleslaw** [comment]
Hi, @avan1235 I'm going to merge tickets related to the Pooling layers during three days and return to these tickets with a review

**Comment by zaleslaw** [comment]
Feel free to create issues and notify me, I'll assign them to you (write a comment about it)

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/util/nameConventions.kt [hunk 1]
```diff
@@ -19,11 +19,11 @@ internal fun defaultInitializerOpName(name: String) = "Init_$name"
...
 
-/** Default Conv2d bias variable name in TensorFlow graph, based on variable's name. */
-internal fun conv2dBiasVarName(name: String) = name + "_" + "conv2d_bias"
+/** Default Conv bias variable name in TensorFlow graph, based on variable's name. */
+internal fun convBiasVarName(name: String, dim: Int) = name + "_" + "conv${dim}d_bias"
 
-/** Default Conv2d kernel variable name in TensorFlow graph, based on variable's name. */
-internal fun conv2dKernelVarName(name: String) = name + "_" + "conv2d_kernel"
+/** Default Conv kernel variable name in TensorFlow graph, based on variable's name. */
+internal fun convKernelVarName(name: String, dim: Int) = name + "_" + "conv${dim}d_kernel"
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/util/nameConventions.kt:26]: Need to ask: is it valid in Keras for conv1d and conv3d? did you check it?

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelLoader.kt [hunk 2]
```diff
@@ -708,6 +709,42 @@ private fun createConv2D(config: LayerConfig, name: String): Conv2D {
     )
 }
 
+private fun createConv3D(config: LayerConfig, name: String): Conv3D {
+    val kernelSize = config.kernel_size!!.map { it.toLong() }.toLongArray()
+    val strides = config.strides!!.map { it.toLong() }.toLongArray()
+
+    val addedOnesStrides = LongArray(5)
+    addedOnesStrides[0] = 1
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelLoader.kt:717]: hmm, seems like we have no chance to serialize all 5d arrays (strides, dilations and so on) https://github.com/JetBrains/KotlinDL/discussions/110 yet one argument to reject them

**Review by zaleslaw — COMMENTED** [review]
Really great PR and refactoring; I suppose it could be done in the future with Separable and Depthwise (if it's theoretically possible and used somewhere)

I suggest waiting until all previous PRs related to pooling will be merged; I'll notify you about it

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/AbstractConv.kt [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+) : Layer(name) {
+    // weight tensors
+    protected lateinit var kernel: Variable<Float>
+    protected var bias: Variable<Float>? = null
+
+    // weight tensor shapes
+    protected lateinit var kernelShape: Shape
+    protected lateinit var biasShape: Shape
+
+    // NDHWC conv3d
```
[the remark this answers is not in the document]
Comment (avan1235) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/AbstractConv.kt:47]: fixed, it was comment to myself

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/AbstractConv.kt [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+import org.tensorflow.Shape
+import org.tensorflow.op.Ops
+import org.tensorflow.op.core.Variable
+import java.lang.IllegalArgumentException
+import kotlin.math.roundToInt
+
+public abstract class AbstractConv(
+    protected val filtersInternal: Long,
+    protected val kernelSizeInternal: LongArray,
+    protected val stridesInternal: LongArray,
```
[the remark this answers is not in the document]
Comment (avan1235) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/AbstractConv.kt:25]: won't fix in this PR but will remember and can refactor in near future

[the remark this answers is not in the document]
Comment (zaleslaw) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/AbstractConv.kt:25]: Of course @avan1235 it's out of scope for this PR

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/Conv2D.kt [hunk 5]
```diff
@@ -102,134 +93,76 @@ public class Conv2D(
...
-        val options: Conv2d.Options = dilations(dilationsInternal.toList()).dataFormat("NHWC")
-        var output: Operand<Float> = tf.nn.conv2d(input, kernel, stridesInternal.toMutableList(), paddingName, options)
-
-        if (useBiasInternal) {
-            output = tf.nn.biasAdd(output, bias)
-        }
-
-        return Activations.convert(activationInternal).apply(tf, output, name)
+        val options = dilations(dilationsInternal.toList()).dataFormat("NHWC")
+        return tf.nn.conv2d(input, kernel, stridesInternal.toMutableList(), paddingInternal.paddingName, options)
```
[the remark this answers is not in the document]
Comment (avan1235) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/convolutional/Conv2D.kt:143]: Could you explain what do you mean in some more details as it is just 2 lines of code function and I don't see any reference to some lambda

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/util/nameConventions.kt (hunk 1 above)
Comment (avan1235) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/util/nameConventions.kt:26]: Tried to create an example but didn't find even `conv2d_kernel`. Could you write me what's wrong [here](https://colab.research.google.com/drive/1EhRfcgtGdb7FwhBIgKnCo095u0hKnO4j?usp=sharing)

Comment (zaleslaw) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/util/nameConventions.kt:26]: Sorry for that, it was the wrong question for you, ignore it.
All these names are used in KotlinDL only for graph manipulations and related to graph only and not related to the Keras imported config (excluding the layer name, of course).
I'm ok with this changes

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelLoader.kt (hunk 2 above)
Comment (zaleslaw) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelLoader.kt:766]: just a comment for me

**Comment by zaleslaw** [comment]
This PR could be merged next week I hope, when all pooling layers will be merged, it takes time to apply some changes in tests

**Comment by zaleslaw** [comment]
1 test failed

org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
  at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
  at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:40)
  at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:35)
  at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:162)
  at org.jetbrains.kotlinx.dl.api.core.models.SequentialModelTest.summary(SequentialCompilationTest.kt:106)

The problem is related to graph building, try to fix this test, it improve your understanding of TF

**Comment by avan1235** [comment]
@zaleslaw that's was great lesson, thanks for leaving that for me 😉 this test now is passed locally so I think you can rerun all TCs

**Comment by avan1235** [comment]
@zaleslaw  I added a sample with sound conv net with the handlers for wav files in kotlinDL so please look to code once again if possible 😉

**Comment by avan1235** [comment]
@zaleslaw I just wanted to point that there is now single test using the introduced dataset so it would be preferable to upload the dataset and let me know to test it locally before running all TCs on your machine because it can potentially fail because of missing data on S3

**Pull request merged**